File: apps/ml/app/image_magic.py
"""File-signature ("magic bytes") sniffing for accepted image MIME types.

Used at the API boundary so we don't trust the client-supplied
``Content-Type`` header — an attacker can label any payload ``image/jpeg``.
"""

from __future__ import annotations

from typing import Final, Literal

AllowedImageMime = Literal["image/jpeg", "image/png", "image/webp"]

ALLOWED_IMAGE_MIMES: Final[frozenset[AllowedImageMime]] = frozenset(
    {"image/jpeg", "image/png", "image/webp"}
)


def sniff_image_mime(buf: bytes) -> AllowedImageMime | None:
    """Return the detected image MIME, or ``None`` if not a supported format."""
    if len(buf) >= 3 and buf[:3] == b"\xff\xd8\xff":
        return "image/jpeg"
    if len(buf) >= 8 and buf[:8] == b"\x89PNG\r\n\x1a\n":
        return "image/png"
    if len(buf) >= 12 and buf[:4] == b"RIFF" and buf[8:12] == b"WEBP":
        return "image/webp"
    return None

File: apps/public/lib/image-magic.ts
/**
 * File-signature ("magic bytes") sniffing for the three image MIME types we
 * accept. Used at the API boundary so we don't trust the client-supplied
 * `Content-Type` header — an attacker can label any payload `image/jpeg`.
 *
 * Returns the detected MIME, or null if the bytes don't match any allowed
 * format.
 */
export type AllowedImageMime = 'image/jpeg' | 'image/png' | 'image/webp';

const PNG_MAGIC = [0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a];

export function sniffImageMime(buf: Uint8Array): AllowedImageMime | null {
  if (buf.length >= 3 && buf[0] === 0xff && buf[1] === 0xd8 && buf[2] === 0xff) {
    return 'image/jpeg';
  }
  if (buf.length >= 8 && PNG_MAGIC.every((b, i) => buf[i] === b)) {
    return 'image/png';
  }
  if (
    buf.length >= 12 &&
    buf[0] === 0x52 &&
    buf[1] === 0x49 &&
    buf[2] === 0x46 &&
    buf[3] === 0x46 &&
    buf[8] === 0x57 &&
    buf[9] === 0x45 &&
    buf[10] === 0x42 &&
    buf[11] === 0x50
  ) {
    return 'image/webp';
  }
  return null;
}
